opal/datatype/opal_datatype_position.c

```diff
@@ -61,22 +61,77 @@ position_predefined_data( opal_convertor_t* CONVERTOR,
                           unsigned char** POINTER,
                           size_t* SPACE )
 {
-    size_t _copy_count = *(COUNT);
-    size_t _copy_blength;
-    ddt_elem_desc_t* _elem = &((ELEM)->elem);
+    const ddt_elem_desc_t* _elem = &((ELEM)->elem);
+    size_t total_count = _elem->count * _elem->blocklen;
+    size_t cando_count = (*SPACE) / opal_datatype_basicDatatypes[_elem->common.type]->size;
+    size_t do_now, do_now_bytes;
+    unsigned char* _memory = (*POINTER) + _elem->disp;
 
-    _copy_blength =  opal_datatype_basicDatatypes[_elem->common.type]->size;
-    if( (_copy_count * _copy_blength) > *(SPACE) ) {
-        _copy_count = *(SPACE) / _copy_blength;
-        if( 0 == _copy_count ) return;  /* nothing to do */
+    assert( *(COUNT) <= _elem->count * _elem->blocklen);
+
+    if( cando_count > *(COUNT) )
+        cando_count = *(COUNT);
+
+    /**
+     * First check if we already did something on this element ?
+     */
+    do_now = (total_count - *(COUNT));  /* done elements */
+    if( 0 != do_now ) {
+        do_now = do_now % _elem->blocklen;  /* partial blocklen? */
+
+        if( 0 != do_now ) {
+            size_t left_in_block = _elem->blocklen - do_now;  /* left in the current blocklen */
+            do_now = (left_in_block > cando_count ) ? cando_count : left_in_block;
+            do_now_bytes = do_now * opal_datatype_basicDatatypes[_elem->common.type]->size;
+
+            OPAL_DATATYPE_SAFEGUARD_POINTER( _memory, do_now_bytes, (CONVERTOR)->pBaseBuf,
+                                            (CONVERTOR)->pDesc, (CONVERTOR)->count );
+            DO_DEBUG( opal_output( 0, "position( %p, %lu ) => space %lu [prolog]\n",
+                                   (void*)_memory, (unsigned long)do_now_bytes, (unsigned long)(*(SPACE)) ); );
+            _memory      = *(POINTER) + _elem->disp + (ptrdiff_t)do_now_bytes;
```

Comment:
_memory += (ptrdiff_t)do_now_bytes
(_memory has already been initialized at the very beginning)